ts/session/sending/PendingMessageCache.ts

```diff
@@ -1,36 +1,149 @@
+    // Get all pending from cache, sorted with oldest first
+    return [...this.cache].sort((a, b) => a.timestamp - b.timestamp);
   }
 
-  public addPendingMessage(
-    device: string,
+  public getForDevice(device: PubKey): Array<RawMessage> {
+    const pending = this.cache.filter(m => m.device === device.key);
+
+    return pending.sort((a, b) => a.timestamp - b.timestamp);
+  }
+
+  public getDevices(): Array<PubKey> {
+    // Gets all unique devices with pending messages
+    const pubkeyStrings = [...new Set(this.cache.map(m => m.device))];
+
+    const pubkeys: Array<PubKey> = [];
+    pubkeyStrings.forEach(pubkey => {
+      if (PubKey.validate(pubkey)) {
+        pubkeys.push(new PubKey(pubkey));
+      }
+    });
+
+    return pubkeys;
+  }
+
+  public async add(
+    device: PubKey,
     message: ContentMessage
-  ): RawMessage {
-    // TODO: Maybe have a util for converting OutgoingContentMessage to RawMessage?
-    // TODO: Raw message has uuid, how are we going to set that? maybe use a different identifier?
-    // One could be device + timestamp would make a unique identifier
-    // TODO: Return previous pending message if it exists
-    return {} as RawMessage;
+  ): Promise<RawMessage> {
+    const rawMessage = MessageUtils.toRawMessage(device, message);
+
+    // Does it exist in cache already?
+    if (this.find(rawMessage)) {
+      return rawMessage;
+    }
+
+    this.cache.push(rawMessage);
+    await this.saveToDB();
+
+    return rawMessage;
   }
 
-  public removePendingMessage(message: RawMessage) {
-    // TODO: implement
+  public async remove(
+    message: RawMessage
+  ): Promise<Array<RawMessage> | undefined> {
+    // Should only be called after message is processed
+
+    // Return if message doesn't exist in cache
+    if (!this.find(message)) {
+      return;
+    }
+
+    // Remove item from cache and sync with database
+    const updatedCache = this.cache.filter(
+      m => m.identifier !== message.identifier
+    );
+    this.cache = updatedCache;
+    await this.saveToDB();
+
+    return updatedCache;
+  }
+
+  public find(message: RawMessage): RawMessage | undefined {
+    // Find a message in the cache
+    return this.cache.find(
+      m => m.device === message.device && m.timestamp === message.timestamp
+    );
+  }
+
+  public async clear() {
+    // Clears the cache and syncs to DB
+    this.cache = [];
+    await this.saveToDB();
   }
 
-  public getPendingDevices(): Array<String> {
-    // TODO: this should return all devices which have pending messages
-    return [];
+  public async loadFromDB() {
+    const messages = await this.getFromStorage();
+    this.cache = messages;
+  }
+
+  private async getFromStorage(): Promise<Array<RawMessage>> {
+    const data = await getItemById('pendingMessages');
+    if (!data || !data.value) {
+      return [];
+    }
+
+    const barePending = JSON.parse(String(data.value));
+
+    const pending = barePending.map((message: any) => {
```

Comment:
I'm assuming you're doing this for the decode of `plainTextBuffer`?

If not you can simplify it down to
```js
const barePending = JSON.parse(String(data.value)) as Array<RawMessage>;
```